### PR TITLE
Laboratory work #3, Diana Gribanova - 22FPL1

### PR DESCRIPTION
**Pull request by Dzianna7**

### CODE REVIEW

**Review by marina-kaz — COMMENTED** [review]
Nothing to check today!

**Review by marina-kaz — COMMENTED** [review]
Nothing to check today :<

File: lab_3_generate_by_ngrams/main.py [hunk 1]
```diff
@@ -41,6 +43,21 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
...
+            return None
+        no_letters = 0
+        for i in text:
+            if not i.isalpha():
+                no_letters += 1
+                if len(text) == no_letters:
+                    return None
+        tokenized_list = []
+        for token in text.lower().split():
+            word = list(''.join(symbol.lower() for symbol in token if symbol.isalpha()))
```
Comment (marina-kaz) [inline lab_3_generate_by_ngrams/main.py:56]: Consider using list comprehension

File: lab_3_generate_by_ngrams/main.py [hunk 2]
```diff
@@ -41,6 +43,21 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
...
         In case any of methods used return None, None is returned.
         """
+        if not isinstance(text, str) or text == '':
+            return None
+        no_letters = 0
+        for i in text:
+            if not i.isalpha():
+                no_letters += 1
+                if len(text) == no_letters:
+                    return None
```
Comment (marina-kaz) [inline lab_3_generate_by_ngrams/main.py:53]: You could have checked it at the very end: if the text consists of digits and spaces, then the resulting tokenized list would be empty, right?

File: lab_3_generate_by_ngrams/main.py [hunk 3]
```diff
@@ -55,6 +72,10 @@ def get_id(self, element: str) -> Optional[int]:
         In case of corrupt input arguments or arguments not included in storage,
         None is returned
         """
+        if not isinstance(element, str) or element not in self._storage:
+            return None
+        if element in self._storage:
```
Comment (marina-kaz) [inline lab_3_generate_by_ngrams/main.py:77]: It was checked before

File: lab_3_generate_by_ngrams/main.py [hunk 4]
```diff
@@ -76,6 +98,13 @@ def get_token(self, element_id: int) -> Optional[str]:
 
         In case of corrupt input arguments or arguments not included in storage, None is returned
         """
+        if not isinstance(element_id, int) or element_id not in self._storage.values():
+            return None
+        appropriate_letter = ''
+        for key, value in self._storage.items():
+            if value == element_id:
+                appropriate_letter = key
+        return appropriate_letter
```
Comment (marina-kaz) [inline lab_3_generate_by_ngrams/main.py:107]: why not return right in the cycle?

File: lab_3_generate_by_ngrams/main.py [hunk 5]
```diff
@@ -93,6 +122,19 @@ def encode(self, text: str) -> Optional[tuple[int, ...]]:
...
+        if not isinstance(text, str):
+            return None
+        tokenized_text = self._tokenize(text)
+        encoded_text = []
+        if tokenized_text is None:
+            return None
+        for token in tokenized_text:
+            self._put(token)
+            token_id = self.get_id(token)
+            if token_id is None:
```
Comment (marina-kaz) [inline lab_3_generate_by_ngrams/main.py:134]: How is it possible?

File: lab_3_generate_by_ngrams/main.py [hunk 6]
```diff
@@ -104,6 +146,11 @@ def _put(self, element: str) -> None:
         In case of corrupt input arguments or invalid argument length,
         an element is not added to storage
         """
+        if not isinstance(element, str) or len(element) != 1:
+            return None
+        if element not in self._storage:
+            self._storage[element] = len(self._storage)
+        return element
```
Comment (marina-kaz) [inline lab_3_generate_by_ngrams/main.py:153]: It is supposed to return `None`

File: lab_3_generate_by_ngrams/main.py [hunk 7]
```diff
@@ -121,6 +168,15 @@ def decode(self, encoded_corpus: tuple[int, ...]) -> Optional[str]:
         In case of corrupt input arguments, None is returned.
         In case any of methods used return None, None is returned.
         """
+        if not isinstance(encoded_corpus, tuple) or encoded_corpus == ():
```
Comment (marina-kaz) [inline lab_3_generate_by_ngrams/main.py:171]: This is not a good way to check if a sequence is empty

File: lab_3_generate_by_ngrams/main.py [hunk 8]
```diff
@@ -121,6 +168,15 @@ def decode(self, encoded_corpus: tuple[int, ...]) -> Optional[str]:
...
+            return None
+        decoded_function = self._decode(encoded_corpus)
```
Comment (marina-kaz) [inline lab_3_generate_by_ngrams/main.py:173]: I have questions for your variable names here: these aren't functions, are they?

File: lab_3_generate_by_ngrams/main.py [hunk 9]
```diff
@@ -159,6 +226,17 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
 
         In case of corrupt input arguments, None is returned
         """
+        if not isinstance(decoded_corpus, tuple) or decoded_corpus == ():
+            return None
+        string_with_small_letters = ""
+        first_capital_letter_str = ""
```
Comment (marina-kaz) [inline lab_3_generate_by_ngrams/main.py:232]: consider using method `capitalize`

File: lab_3_generate_by_ngrams/main.py [hunk 10]
```diff
@@ -159,6 +226,17 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
...
+        for letters in decoded_corpus:
+            string_with_small_letters += letters
+            first_capital_letter_str = string_with_small_letters[0].upper() + string_with_small_letters[1:]
+        if self._end_of_word_token in first_capital_letter_str:
+            sentence_with_gaps = first_capital_letter_str.replace(self._end_of_word_token, " ")
+            final_text = sentence_with_gaps[:-1] + "."
+            return final_text
```
Comment (marina-kaz) [inline lab_3_generate_by_ngrams/main.py:239]: is this the right place for return?

**Review by marina-kaz — COMMENTED** [review]
Good progress, keep going!

File: lab_3_generate_by_ngrams/main.py (hunk 3 above)
Comment (Dzianna7) [reply lab_3_generate_by_ngrams/main.py:77]: done

File: lab_3_generate_by_ngrams/main.py (hunk 4 above)
Comment (Dzianna7) [reply lab_3_generate_by_ngrams/main.py:107]: done

File: lab_3_generate_by_ngrams/main.py (hunk 6 above)
Comment (Dzianna7) [reply lab_3_generate_by_ngrams/main.py:153]: done

File: lab_3_generate_by_ngrams/main.py (hunk 7 above)
Comment (Dzianna7) [reply lab_3_generate_by_ngrams/main.py:171]: done

File: lab_3_generate_by_ngrams/main.py (hunk 8 above)
Comment (Dzianna7) [reply lab_3_generate_by_ngrams/main.py:173]: i meant only the fact that in this variable i will save the work of the _decode function
Should i check the names?

File: lab_3_generate_by_ngrams/main.py (hunk 9 above)
Comment (Dzianna7) [reply lab_3_generate_by_ngrams/main.py:232]: done

File: lab_3_generate_by_ngrams/main.py (hunk 10 above)
Comment (Dzianna7) [reply lab_3_generate_by_ngrams/main.py:239]: done

**Review by marina-kaz — COMMENTED** [review]
I see the progress, but I do not see any changes concerning my comments

**Review by marina-kaz — COMMENTED** [review]
No changes since my last review

**Pull request closed without merging**